### PR TITLE
Biblio search completer

### PR DESCRIPTION
**Pull request by alinadubatovka**

### CODE REVIEW

**Comment by dbarashev** [comment]
@khazhoyan Arsen, can you look at the custom actions in biblio autocompletion?

File: lib/ace/ext/papeeria/biblio_completer.coffee [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+  class BiblioFilesCompleter extends Autocomplete.Autocomplete
+    constructor: ->
+      super()
+    setFiles: (files) => @files = files
+    _setCompletions: =>
+      matches = @files.map((file) -> return {
+        name: file.name
+        file: file
+        meta: file.path
+        completer: this
```
Comment (khazhoyan) [inline lib/ace/ext/papeeria/biblio_completer.coffee]: is it true that this is used only in the `Autocomplete#insertMatch` method? which btw is overridden in this class so it seems unnecessary to me

File: lib/ace/ext/papeeria/biblio_completer.coffee [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+      return
+
+    # Will be called when users choose a Search option
+    setSearchCallback: (callback) => @searchCallback = callback
+    # Will be called when users choose an Import option
+    setImportCallback: (callback) => @importCallback = callback
+    insertMatch: (editor, data) =>
+      editor.completer.detach()
+      editor.completer = null
+      if data.name == "Search"
```
Comment (khazhoyan) [inline lib/ace/ext/papeeria/biblio_completer.coffee]: the same about carrying callback with the data

File: lib/ace/ext/papeeria/biblio_completer.coffee [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+      callback(null, [])
+      return
+
+    # Will be called when users choose a Search option
+    setSearchCallback: (callback) => @searchCallback = callback
+    # Will be called when users choose an Import option
+    setImportCallback: (callback) => @importCallback = callback
+    insertMatch: (editor, data) =>
+      editor.completer.detach()
```
Comment (khazhoyan) [inline lib/ace/ext/papeeria/biblio_completer.coffee]: the same about `@isDisposable = true`

File: lib/ace/ext/papeeria/biblio_completer.coffee [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+define((require, exports, module) ->
+  LatexParsingContext = require("ace/ext/papeeria/latex_parsing_context")
+  Autocomplete = require('ace/autocomplete')
+  util = require("ace/autocomplete/util")
+  lang = require("../../lib/lang")
+
+  class BiblioFilesCompleter extends Autocomplete.Autocomplete
+    constructor: ->
+      super()
+    setFiles: (files) => @files = files
```
Comment (khazhoyan) [inline lib/ace/ext/papeeria/biblio_completer.coffee]: if you don't mind, please separate method with blank lines

File: lib/ace/ext/papeeria/biblio_completer.coffee [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    _setCompletions: =>
+      matches = @files.map((file) -> return {
```
Comment (khazhoyan) [inline lib/ace/ext/papeeria/biblio_completer.coffee]: FYI: `return` isn't necessary in one-statement lambdas

File: lib/ace/ext/papeeria/biblio_completer.coffee [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+    constructor: () ->
+      @enabledMendeley = false
+
+    setEnabledMendeley: (enabledMendeley) => @enabledMendeley = enabledMendeley
+
+    getCompletions: (editor, session, pos, prefix, callback) =>
+      token = session.getTokenAt(pos.row, pos.column)
+
+      if LatexParsingContext.isType(token, "cite")
+        default_result = [{
```
Comment (khazhoyan) [inline lib/ace/ext/papeeria/biblio_completer.coffee]: I believe we're more accustomed to camelCase in coffee code :)

File: lib/ace/ext/papeeria/biblio_completer.coffee [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+        @detach()
+        return
+
+      if data.name == "Create new file"
+        @_doCreate()
+      else
+        @_doInsert(data.file)
+
+    detach: =>
+      @editor.completer = new Autocomplete.Autocomplete()
```
Comment (khazhoyan) [inline lib/ace/ext/papeeria/biblio_completer.coffee]: try adding `@isDisposable = true` field (like it's done in `spellchecker_popup.coffee`) and chances are you won't need to override `detach`

File: lib/ace/ext/papeeria/biblio_completer.coffee [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+    setFiles: (files) => @files = files
+    _setCompletions: =>
+      matches = @files.map((file) -> return {
+        name: file.name
+        file: file
+        meta: file.path
+        completer: this
+      })
+      matches.push({
+        name: "Create new file"
```
Comment (alinadubatovka) [inline lib/ace/ext/papeeria/biblio_completer.coffee]: Done

File: lib/ace/ext/papeeria/biblio_completer.coffee [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+      @openPopup(@editor, prefix, keepPopupPosition)
+
+    insertMatch: (editor, data) =>
+      if not data?
+        data = @popup.getData(@popup.getRow())
+      if not data?
+        @detach()
+        return
+
+      if data.name == "Create new file"
```
Comment (alinadubatovka) [inline lib/ace/ext/papeeria/biblio_completer.coffee]: Cool, thanks

File: lib/ace/ext/papeeria/biblio_completer.coffee [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+      @enabledMendeley = false
+
+    setEnabledMendeley: (enabledMendeley) => @enabledMendeley = enabledMendeley
+
+    getCompletions: (editor, session, pos, prefix, callback) =>
+      token = session.getTokenAt(pos.row, pos.column)
+
+      if LatexParsingContext.isType(token, "cite")
+        default_result = [{
+          name: "Search"
```
Comment (alinadubatovka) [inline lib/ace/ext/papeeria/biblio_completer.coffee]: Done

File: lib/ace/ext/papeeria/biblio_completer.coffee [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+        @_doCreate()
+      else
+        @_doInsert(data.file)
+
+    detach: =>
+      @editor.completer = new Autocomplete.Autocomplete()
+      super()
+
+
+  class BiblioCompleter
```
Comment (alinadubatovka) [inline lib/ace/ext/papeeria/biblio_completer.coffee]: As far as I understand, no. It just provides Search and Import options to Autocomplete, which also collects completions from tex, text, keyboard and snippet completers and show them all together.

File: lib/ace/ext/papeeria/biblio_completer.coffee [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+    constructor: () ->
+      @enabledMendeley = false
+
+    setEnabledMendeley: (enabledMendeley) => @enabledMendeley = enabledMendeley
```
Comment (alinadubatovka) [inline lib/ace/ext/papeeria/biblio_completer.coffee]: Fixed

File: lib/ace/ext/papeeria/biblio_completer.coffee (hunk 4 above)
Comment (alinadubatovka) [inline lib/ace/ext/papeeria/biblio_completer.coffee]: Sorry, fixed

File: lib/ace/ext/papeeria/biblio_completer.coffee (hunk 6 above)
Comment (alinadubatovka) [inline lib/ace/ext/papeeria/biblio_completer.coffee]: :(

Comment (alinadubatovka) [inline lib/ace/ext/papeeria/biblio_completer.coffee]: Fixed

File: lib/ace/ext/papeeria/biblio_completer.coffee (hunk 2 above)
Comment (alinadubatovka) [inline lib/ace/ext/papeeria/biblio_completer.coffee]: Done

File: lib/ace/ext/papeeria/biblio_completer.coffee [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+        completer: this
+      })
+      matches.push({
+        name: "Create new file"
+        meta: ""
+        completer: this
+      })
+      @completions = new Autocomplete.FilteredList(matches)
+
+    setCreateCallback: (callback) => @createCallback = callback
```
Comment (alinadubatovka) [inline lib/ace/ext/papeeria/biblio_completer.coffee]: Fixed

File: lib/ace/ext/papeeria/biblio_completer.coffee (hunk 3 above)
Comment (alinadubatovka) [inline lib/ace/ext/papeeria/biblio_completer.coffee]: `BiblioCompleter` does not extend `Autocomplete.Autocomplete`, so it seems useless (and doesn't work, I've tried :)

File: lib/ace/ext/papeeria/biblio_completer.coffee (hunk 7 above)
Comment (alinadubatovka) [inline lib/ace/ext/papeeria/biblio_completer.coffee]: For some reason it doesn't work, but thanks anyway

File: lib/ace/ext/papeeria/biblio_completer.coffee (hunk 5 above)
Comment (alinadubatovka) [inline lib/ace/ext/papeeria/biblio_completer.coffee]: Fixed, thanks

File: lib/ace/ext/papeeria/biblio_completer.coffee [hunk 14]
```diff
@@ -39,76 +37,71 @@ define((require, exports, module) ->
...
-        default_result = [{
-          name: "Search"
-          snippet: "Search " + prefix
+        defaultResult = [{
+          name: @searchLabel
+          snippet: @searchLabel + prefix
           meta: ""
           score: 1000
           meta_score: 10
           completer: this
```
Comment (khazhoyan) [inline lib/ace/ext/papeeria/biblio_completer.coffee]: maybe remove this `completer: this` as well?

File: lib/ace/ext/papeeria/biblio_completer.coffee [hunk 15]
```diff
@@ -39,76 +37,71 @@ define((require, exports, module) ->
...
-          default_result.push({
-            name: "Import"
-            snippet: "Import " + prefix
-            meta: "from Mendeley"
+        if @enabledMendeley?
+          defaultResult.push({
+            name: @importLabel
+            snippet: @importLabel + prefix
+            meta: @importSource
             score: 1000
```
Comment (khazhoyan) [inline lib/ace/ext/papeeria/biblio_completer.coffee]: you sure a thousand is enough? I used `Number.MAX_VALUE`. 
and please extract it to a constant :)
also make sure it's okay that both import and search objects have the same `score`

File: lib/ace/ext/papeeria/biblio_completer.coffee [hunk 16]
```diff
@@ -1,31 +1,29 @@
...
 define((require, exports, module) ->
   LatexParsingContext = require("ace/ext/papeeria/latex_parsing_context")
-  Autocomplete = require('ace/autocomplete')
+  Autocomplete = require("ace/autocomplete")
   util = require("ace/autocomplete/util")
   lang = require("../../lib/lang")
 
   class BiblioFilesCompleter extends Autocomplete.Autocomplete
-    constructor: ->
+    constructor:(@createCallback, @insertCallback, @createLabel) ->
```
Comment (khazhoyan) [inline lib/ace/ext/papeeria/biblio_completer.coffee]: space before the `(`?

File: lib/ace/ext/papeeria/biblio_completer.coffee [hunk 17]
```diff
@@ -39,76 +37,71 @@ define((require, exports, module) ->
...
       super()
 
 
   class BiblioCompleter
-    constructor: () ->
-      @enabledMendeley = false
-
-    setEnabledMendeley: (enabledMendeley) => @enabledMendeley = enabledMendeley
+    constructor: (@enabledMendeley, @searchLabel, @importLabel, @importSource, @searchCallback) ->
+      @isDisposable = true
```
Comment (khazhoyan) [inline lib/ace/ext/papeeria/biblio_completer.coffee]: remove this if this doesn't help? :)

File: lib/ace/ext/papeeria/biblio_completer.coffee [hunk 18]
```diff
@@ -39,76 +37,71 @@ define((require, exports, module) ->
...
       @editor.completer = new Autocomplete.Autocomplete()
       super()
 
 
   class BiblioCompleter
-    constructor: () ->
-      @enabledMendeley = false
-
-    setEnabledMendeley: (enabledMendeley) => @enabledMendeley = enabledMendeley
+    constructor: (@enabledMendeley, @searchLabel, @importLabel, @importSource, @searchCallback) ->
```
Comment (khazhoyan) [inline lib/ace/ext/papeeria/biblio_completer.coffee]: maybe `@isMendeleyEnabled` would be better for a boolean?

File: lib/ace/ext/papeeria/biblio_completer.coffee (hunk 7 above)
Comment (khazhoyan) [inline lib/ace/ext/papeeria/biblio_completer.coffee]: alright. btw is creating a new `Autocomplete` instance is necessary here? doesn't just `@editor.completer = null` work?

File: lib/ace/ext/papeeria/biblio_completer.coffee (hunk 14 above)
Comment (alinadubatovka) [inline lib/ace/ext/papeeria/biblio_completer.coffee]: Then it would be call `Autocomplete#insertMatch`

File: lib/ace/ext/papeeria/biblio_completer.coffee (hunk 16 above)
Comment (alinadubatovka) [inline lib/ace/ext/papeeria/biblio_completer.coffee]: yes, thanks

File: lib/ace/ext/papeeria/biblio_completer.coffee (hunk 14 above)
Comment (khazhoyan) [inline lib/ace/ext/papeeria/biblio_completer.coffee]: for the record: pls add a comment explaining this

File: lib/ace/ext/papeeria/biblio_completer.coffee (hunk 17 above)
Comment (alinadubatovka) [inline lib/ace/ext/papeeria/biblio_completer.coffee]: Done

File: lib/ace/ext/papeeria/biblio_completer.coffee (hunk 18 above)
Comment (alinadubatovka) [inline lib/ace/ext/papeeria/biblio_completer.coffee]: Okay, renamed
